### PR TITLE
Do not overwrite direct link titles with redirecting titles

### PR DESCRIPTION
**Pull request by ondenman**
Fixes: https://github.com/everypolitician/wikidata-fetcher/issues/44 and https://github.com/everypolitician/wikidata-fetcher/issues/46

The wikipedia page for 'The Deeper Meaning of Liff' redirects to 'The Meaning of Life'.

`WikiData.ids_from_pages('en', ['The Meaning of Liff', 'The Deeper Meaning of Liff'])` currently only returns:

```{'The Deeper Meaning of Liff' => 'Q875382'}```

No wikidata ID is returned of 'The Meaning of Liff'.

This commit returns the wikidata id with both links:
```
{ 'The Meaning of Liff' => 'Q875382', 'The Deeper Meaning of Liff' => 'Q875382' }
```

### CODE REVIEW

File: lib/wikidata.rb [hunk 1]
```diff
@@ -19,10 +19,16 @@ def self.ids_from_pages(lang, titles)
         token_type: false,
       }
       response = client.action :query, page_args
-      redirected_from = Hash[(response.data['redirects'] || []).map { |h| [h['to'], h['from']] }]
-      response.data['pages'].select { |_k, v| v.key? 'pageprops' }.map do |_k, v|
-        [redirected_from[v['title']] || v['title'], v['pageprops']['wikibase_item']]
+      redirected_from = Hash[(response.data['redirects'] || []).map { |h| [h['from'], h['to']] }]
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: You're no longer _using_ this as a Hash, but instead just iterating over it, so there's no real need to cast it to one.

File: lib/wikidata.rb [hunk 2]
```diff
@@ -19,10 +19,16 @@ def self.ids_from_pages(lang, titles)
...
+      titles_and_ids = response.data['pages'].select { |_k, v| v.key? 'pageprops' }.map do |_k, v|
+        [v['title'], v['pageprops']['wikibase_item']]
       end
+      redirect_titles_and_ids = redirected_from.map do |k, v|
+        title_and_id = titles_and_ids.find { |a| a.include? v }
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: This is odd on two fronts: 

* firstly, the `a.include?` is also at a very odd level of specificity, that disguises what you're actually testing. (i.e. that the first element of the Array == `v`).
* but secondly, you really shouldn't be calling `find` against an Array in a loop anyway — it would be much better to convert the original array to a Hash to check against.

File: t/redirects.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+require 'test_helper'
+
+describe 'redirecting links' do
+  around { |test| VCR.use_cassette('liff', &test) }
+  subject { WikiData.ids_from_pages('en', ['The Meaning of Liff', 'The Deeper Meaning of Liff']) }
```
Comment (tmtmtmtm) [inline t/redirects.rb]: This is a useful test, but I worry that it's a little too lightweight — I can think of all sorts of implementations that could make _this_ code pass, but fail in other scenarios.

I'd suggest leaving this one as is, but also adding a more robust version, that looks up quite a few other pages including at least one that has no redirects in either direction, one where you have a redirect to a page that you haven't specified in your list, and a combination where at least 2 or more different pages all redirect to the same page.

Comment (ondenman) [inline t/redirects.rb]: @tmtmtmtm I am going to handle multiple redirects in another PR as it's a separate issue (https://github.com/everypolitician/wikidata-fetcher/issues/46). (I added the change in the first commit but I've removed it in the fixup.) Also, I haven't yet added a test for a page without incoming or outgoing redirects as I'm not quite sure how to find one.

Comment (tmtmtmtm) [inline t/redirects.rb]: I don't mind if you do things in separate PRs, but I don't think this could go live like this.

Comment (ondenman) [inline t/redirects.rb]: Multiple redirects to a common article are handled in the first commit (3f7cfe3)

The part that swaps the `from` and `to`:
``redirected_from = (response.data['redirects'] || []).map { |h| [h['from'], h['to']] }``

The change is not separate to the change the commit message describes.

I should make the change separately but I'm not sure what the best thing to do would be: should I split the commit and keep both changes in this PR or should I revert the change in a fix up and make it in a separate PR that targets this branch?

Comment (ondenman) [inline t/redirects.rb]: It actually looks like multiple redirects come free with the first change anyway without changing existing code.

File: t/redirects.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    let(:titles) { ['Wikkit Gate', 'Infinite Improbability Drive'] }
+
+    it 'returns wikidata ids paired with redirecting links and direct links' do
+      subject.must_equal("Technology in The Hitchhiker's Guide to the Galaxy" => 'Q259299',
+                         'Infinite Improbability Drive'                       => 'Q259299',
+                         'Wikkit Gate'                                        => 'Q259299')
+    end
+  end
+
+  describe 'page without redirects' do
```
Comment (tmtmtmtm) [inline t/redirects.rb]: The ordering here makes this quite hard to follow. In general it's best to think of the code being read linearly — i.e. you should start with the simplest cases, and built up to the more complex ones.

File: t/redirects.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  describe 'redirect and direct link' do
+    let(:titles) { ['The Meaning of Liff', 'The Deeper Meaning of Liff'] }
+
+    it 'pairs both redirecting link and direct link with wikidata ID' do
+      subject.must_equal('The Meaning of Liff' => 'Q875382', 'The Deeper Meaning of Liff' => 'Q875382')
+    end
+  end
+
+  describe 'only redirect links' do
```
Comment (tmtmtmtm) [inline t/redirects.rb]: it's a little hard to follow what specifically this is testing, in part because the simpler case of a single item that is a redirect doesn't have its own test. Is this meant to _be_ that case? (and if so, why test two?) Or is this testing something more than that? (in which case, it would probably be better to have the simpler case separately, and make it more obvious what this one is testing)

File: t/redirects.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    let(:titles) { ['Marvin the Paranoid Android', 'The Deeper Meaning of Liff'] }
+
+    it 'returns wikidata ids paired with redirecting links and direct links' do
+      subject.must_equal('Marvin (character)'          => 'Q264685',
+                         'The Meaning of Liff'         => 'Q875382',
+                         'The Deeper Meaning of Liff'  => 'Q875382',
+                         'Marvin the Paranoid Android' => 'Q264685')
```
Comment (tmtmtmtm) [inline t/redirects.rb]: This result seems slightly unexpected. If I'm looking up those two items, why are they not the only two I'm getting back?

File: lib/wikidata.rb [hunk 7]
```diff
@@ -19,10 +19,16 @@ def self.ids_from_pages(lang, titles)
         token_type: false,
       }
       response = client.action :query, page_args
-      redirected_from = Hash[(response.data['redirects'] || []).map { |h| [h['to'], h['from']] }]
-      response.data['pages'].select { |_k, v| v.key? 'pageprops' }.map do |_k, v|
-        [redirected_from[v['title']] || v['title'], v['pageprops']['wikibase_item']]
+
```
Comment (ondenman) [inline lib/wikidata.rb]: @tmtmtmtm I've taken a stab at breaking it out into its own class (fix up: 592a14c
) but I'm not sure if I've headed in the right direction or not. I suspect the 'and' in the class name is telling me I haven't.

**Comment by ondenman** [comment]
Rubocop fails after `6370046`. Some of the failing code is pre-existing (eg: `lib/wikidata.rb:69:5: E: dynamic constant assignment`). Updating Rubocop target to 2.3 fixes it but I don't yet know why it passes previous to the commit.

File: lib/wikidata.rb [hunk 8]
```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
...
+    end
+
+    def all
+      # response.data['pages'] contains the direct title of the article
+      # regardless of whether the query contains the direct title or not. To
+      # ensure we return only the titles requested, we map the results to the
+      # titles provided, so that any unrequested direct titles are not
+      # included in the result.
+      titles.reduce([]) do |arr, title|
+        arr << [title, titles_hash[title]] unless titles_hash[title].nil?
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: `reduce` is almost always going to be the wrong approach to this sort of thing. 

I'm not entirely sure I understand what's going on here, but this seems to be simpler and clearer way of doing the same thing

```ruby
titles.reject { |t| titles_hash[t].nil? }.map {|t| [t, titles_hash[t]] }
```

File: lib/wikidata.rb [hunk 9]
```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
...
-      end
+      TitlesAndIds.new(titles, client.action(:query, page_args)).all
     end
     results = Hash[res.flatten(1)]
     missing = titles - results.keys
     warn "Can't find Wikidata IDs for: #{missing.join(', ')} in #{lang}" if missing.any?
     results
   end
+
+  class TitlesAndIds
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: I'm afraid I still don't really understand what this class is or does. I think you need to document this.

File: lib/wikidata.rb [hunk 10]
```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
...
+      # titles provided, so that any unrequested direct titles are not
+      # included in the result.
+      titles.reduce([]) do |arr, title|
+        arr << [title, titles_hash[title]] unless titles_hash[title].nil?
+      end
+    end
+
+    private
+
+    attr_reader :response, :titles
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: I'm surprised `rubocop` doesn't insist on a blank line after this. Or did you have one and it took it out again?

File: lib/wikidata.rb [hunk 11]
```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
...
+    end
+
+    def direct_titles
+      response.data['pages'].values.select { |v| v.key? 'pageprops' }.map do |v|
+        [v['title'], v['pageprops']['wikibase_item']]
+      end
+    end
+
+    def redirect_titles
+      (response.data['redirects'] || []).map do |h|
```
Comment (tmtmtmtm) [inline lib/wikidata.rb]: I think you can just `return [] unless `response.data['redirect']` here, to make this clearer and simpler, but there doesn't seem to be a test case for that scenario to verify this.

File: t/ids_from_pages.rb [hunk 12]
```diff
@@ -27,4 +27,40 @@
...
+                           'Marvin the Paranoid Android' => 'Q264685')
+      end
+    end
+
+    describe 'multiple links redirecting to the same article' do
+      let(:titles) { ['Wikkit Gate', 'Infinite Improbability Drive'] }
+
+      it 'returns wikidata ids paired with redirecting links and direct links' do
+        subject.must_equal('Wikkit Gate'                  => 'Q259299',
+                           'Infinite Improbability Drive' => 'Q259299')
```
Comment (tmtmtmtm) [inline t/ids_from_pages.rb]: The implementation here seems a little too lossy — this is throwing away that these are redirecting to "Technology in The Hitchhiker's Guide to the Galaxy"
